MIGRATION_GUIDE.md

```diff
@@ -450,150 +454,162 @@ The Notifications namespace is accessible via `OneSignal.Notifications` and prov
+    OneSignal.Notifications.removePermissionObserver(self as OSNotificationPermissionObserver)
 ```
 
-### Notification Will Show in Foreground Handler
+### Notification Foreground Lifecycle Listener
+Any object implementing the `OSNotificationLifecycleListener` protocol can be added as a listener. You can call `removeForegroundLifecycleListener` to remove any existing listeners.
 
 **Objective-C**
 ```objc
-    id notificationWillShowInForegroundBlock = ^(OSNotification *notification, OSNotificationDisplayResponse completion) {
-        NSLog(@"Received Notification - %@", notification.notificationId);
-        if ([notification.notificationId isEqualToString:@"silent_notif"]) {
-            completion(nil);
-        } else {
-            completion(notification);
+    // AppDelegate.h
+    // Add OSNotificationLifecycleListener after UIApplicationDelegate
+    @interface AppDelegate : UIResponder <UIApplicationDelegate, OSNotificationLifecycleListener>
+    @end
+
+    // AppDelegate.m
+    @implementation AppDelegate
+
+    - (BOOL)application:(UIApplication*)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
+      // Add your AppDelegate as an observer
+      [OneSignal.Notifications addForegroundLifecycleListener:self];
+    }
+
+    // Add this new method
+
+    - (void)onWillDisplayNotification:(OSNotificationWillDisplayEvent *)event {
+        NSLog(@"Received Notification - %@", event.notification.notificationId);
+        if ([event.notification.notificationId isEqualToString:@"silent_notif"]) {
+            [event preventDefault];
         }
-    };
-    
-    [OneSignal.Notifications setNotificationWillShowInForegroundHandler:notificationWillShowInForegroundBlock];
+
+        // If you called preventDefault, you can call display within 25 seconds
+        [event.notification display];
+    }
+
+    @end
+
+    // Remove the observer
+    [OneSignal.Notifications removeForegroundLifecycleListener:self];
 ```
 **Swift**
 ```swift
-    let notificationWillShowInForegroundBlock: OSNotificationWillShowInForegroundBlock = { notification, completion in
-        print("Received Notification: ", notification.notificationId ?? "no id")
-        print("launchURL: ", notification.launchURL ?? "no launch url")
-        print("content_available = \(notification.contentAvailable)")
-    
-        if notification.notificationId == "example_silent_notif" {
-            // Complete with null means don't show a notification
-            completion(nil)
-        } else {
-            // Complete with a notification means it will show
-            completion(notification)
-        } 
+class MyNotificationLifecycleListener : NSObject, OSNotificationLifecycleListener {
+    func onWillDisplay(event: OSNotificationWillDisplayEvent) {
+        // Example of conditionally displaying a notification
+        if event.notification.notificationId == "example_silent_notif" {
```

Comment:
A better example would be looking at `additionalData` instead of `notificationId`.